File: backend/src/schemas/blank_stock.py
"""Pydantic schemas for the Blank Pieces (peças lisas) feature.

A ``BlankPiece`` is a print-agnostic blank garment body keyed by
``(spec, size, color_code)``. On-hand is derived live from a single
append-only ledger (``blank_piece_movements``): ENTRY and ADJUSTMENT credit
stock, EXIT debits it. The service aggregates ledger rows into
``BlankPieceLevelRead`` per blank piece (every catalog row is surfaced, even
with zero movements); movement mutations are POST-only and surface as
``BlankMovementRead``.

Quantities are integer counts (not Decimal like the metered rolls). ``low_stock``
is computed per row from ``min_stock`` (when set) else the company-wide
``blank`` threshold; ``in_production`` is a typed placeholder (0) until the WIP
wiring lands in a later phase.
"""

from __future__ import annotations

import uuid
from datetime import date, datetime

from pydantic import BaseModel, Field

from models.enums import BlankMovementKind, Size
from schemas._common import Page

# ---------- Catalog create ----------


class BlankPieceCreate(BaseModel):
    spec_id: uuid.UUID
    size: Size
    color: str = Field(min_length=1, max_length=40)
    color_code: str = Field(pattern=r"^[A-Z]{3}$")
    min_stock: int | None = Field(default=None, ge=0)


# ---------- Nested DTOs ----------


class SpecMini(BaseModel):
    id: uuid.UUID
    code: str
    name: str


class BlankPieceMini(BaseModel):
    id: uuid.UUID
    spec_code: str
    size: Size
    color: str


# ---------- Blank piece x on-hand aggregate ----------


class BlankPieceLevelRead(BaseModel):
    """A single blank piece's aggregated on-hand position."""

    blank_piece_id: uuid.UUID
    spec_id: uuid.UUID
    spec: SpecMini
    size: Size
    color: str
    color_code: str
    min_stock: int | None = None
    on_hand: int
    in_production: int
    low_stock: bool
    entries_total: int
    exits_total: int
    last_movement_at: datetime | None = None


class BlankPieceLevelSummary(BaseModel):
    """Tenant-wide headline figures for the Peças Lisas page (all SKUs, not a page)."""

    total_on_hand: int
    below_min: int
    sku_count: int


# ---------- Ledger movements ----------


class BlankMovementCreate(BaseModel):
    blank_piece_id: uuid.UUID
    kind: BlankMovementKind = Field(default=BlankMovementKind.ENTRY)
    quantity: int = Field(gt=0)
    notes: str | None = Field(default=None, max_length=500)


class BlankMovementRead(BaseModel):
    id: uuid.UUID
    blank_piece_id: uuid.UUID
    blank_piece: BlankPieceMini | None = None
    kind: BlankMovementKind
    quantity: int
    notes: str | None
    created_at: datetime


# ---------- Filters ----------


class BlankPieceLevelFilters(BaseModel):
    q: str | None = Field(default=None, max_length=120)
    spec_id: uuid.UUID | None = None
    size: Size | None = None
    low_stock_only: bool = False


class BlankMovementFilters(BaseModel):
    blank_piece_id: uuid.UUID | None = None
    kind: BlankMovementKind | None = None
    date_from: date | None = None
    date_to: date | None = None


# ---------- Page aliases ----------


BlankPieceLevelPage = Page[BlankPieceLevelRead]
BlankMovementPage = Page[BlankMovementRead]

File: frontend/src/lib/schemas/blank-stock.ts
/**
 * Zod schemas for the Blank Pieces (peças lisas) tier.
 *
 * Mirrors the Pydantic shapes from `backend/src/schemas/blank_stock.py`. A
 * `BlankPiece` is a print-agnostic garment body keyed by
 * `(spec, size, color_code)`. On-hand is derived live from an append-only
 * ledger (`blank_piece_movements`): ENTRY and ADJUSTMENT credit, EXIT debits.
 * Quantities are integer counts (not Decimal like the metered rolls), so
 * `on_hand`/`in_production`/`quantity` are numbers and `low_stock` a boolean.
 */

import { z } from "zod";
import { SIZES } from "@/lib/schemas/product";

export const BLANK_MOVEMENT_KINDS = ["entry", "exit", "adjustment"] as const;
export const blankMovementKindSchema = z.enum(BLANK_MOVEMENT_KINDS);
export type BlankMovementKind = z.infer<typeof blankMovementKindSchema>;

const pageMeta = {
  total: z.number(),
  page: z.number(),
  page_size: z.number(),
  has_more: z.boolean(),
};

// ---------- Nested DTOs ----------

export const specMiniSchema = z.object({
  id: z.string(),
  code: z.string(),
  name: z.string(),
});
export type SpecMini = z.infer<typeof specMiniSchema>;

export const blankPieceMiniSchema = z.object({
  id: z.string(),
  spec_code: z.string(),
  size: z.enum(SIZES),
  color: z.string(),
});
export type BlankPieceMini = z.infer<typeof blankPieceMiniSchema>;

// ---------- Levels (blank piece x on-hand) ----------

export const blankPieceLevelReadSchema = z.object({
  blank_piece_id: z.string(),
  spec_id: z.string(),
  spec: specMiniSchema,
  size: z.enum(SIZES),
  color: z.string(),
  color_code: z.string(),
  min_stock: z.number().int().nullable().optional(),
  on_hand: z.number().int(),
  in_production: z.number().int(),
  low_stock: z.boolean(),
  entries_total: z.number().int(),
  exits_total: z.number().int(),
  last_movement_at: z.string().nullable().optional(),
});
export type BlankPieceLevelRead = z.infer<typeof blankPieceLevelReadSchema>;

export const blankPieceLevelPageSchema = z.object({
  items: z.array(blankPieceLevelReadSchema),
  ...pageMeta,
});
export type BlankPieceLevelPage = z.infer<typeof blankPieceLevelPageSchema>;

// Tenant-wide headline totals (every SKU, not the current page) for the KPIs.
export const blankPieceLevelSummarySchema = z.object({
  total_on_hand: z.number().int(),
  below_min: z.number().int(),
  sku_count: z.number().int(),
});
export type BlankPieceLevelSummary = z.infer<typeof blankPieceLevelSummarySchema>;

// ---------- Movements ledger ----------

export const blankMovementReadSchema = z.object({
  id: z.string(),
  blank_piece_id: z.string(),
  blank_piece: blankPieceMiniSchema.nullable().optional(),
  kind: blankMovementKindSchema,
  quantity: z.number().int(),
  notes: z.string().nullable(),
  created_at: z.string(),
});
export type BlankMovementRead = z.infer<typeof blankMovementReadSchema>;

export const blankMovementPageSchema = z.object({
  items: z.array(blankMovementReadSchema),
  ...pageMeta,
});
export type BlankMovementPage = z.infer<typeof blankMovementPageSchema>;

// ---------- Create payloads ----------

export type BlankPieceCreate = {
  spec_id: string;
  size: (typeof SIZES)[number];
  color: string;
  color_code: string;
  min_stock?: number | null;
};

export type BlankMovementCreate = {
  blank_piece_id: string;
  kind: BlankMovementKind;
  quantity: number;
  notes?: string | null;
};

// ---------- Filters ----------

export type BlankPieceLevelFilters = {
  q?: string;
  spec_id?: string;
  size?: (typeof SIZES)[number];
  low_stock_only?: boolean;
  page?: number;
  page_size?: number;
};

export type BlankMovementFilters = {
  blank_piece_id?: string;
  kind?: BlankMovementKind;
  date_from?: string;
  date_to?: string;
  page?: number;
  page_size?: number;
};
